File: simsys_metrics/_process.py
"""Custom Prometheus collector for simsys_process_* metrics.

``prometheus_client`` ships a ProcessCollector by default but its metric names
don't carry the ``simsys_`` prefix and they don't have a ``service`` label.
Rather than re-wrap those, we emit our own via psutil in a single Collector.

Registered once per process by ``install()``. Collectors are idempotent to
register: calling ``register_process_collector(service)`` again with the same
service reuses the existing collector.
"""

from __future__ import annotations

from dataclasses import dataclass
from threading import Lock
from typing import Optional

import psutil
from prometheus_client import REGISTRY
from prometheus_client.metrics_core import CounterMetricFamily, GaugeMetricFamily

_lock = Lock()
_collector: Optional["SimsysProcessCollector"] = None


@dataclass(frozen=True)
class ProcessCollectorRollbackState:
    """Snapshot of the process-collector singleton immediately before
    register_process_collector() mutated it.

    Adapter rollback passes this back to restore_process_collector() so
    a service-swap-then-fail sequence doesn't leave a prior install's
    process metrics broken (entire collector unregistered) and a fresh
    registration doesn't leave dangling metrics in the registry.

    ``action`` describes what register_process_collector did:
      * ``"reused"``       — singleton already had this service; nothing changed.
      * ``"registered"``   — no collector existed; we registered fresh. Rollback drops it.
      * ``"service_swap"`` — collector existed but for a different service; we
                              unregistered it, registered a new one. Rollback
                              drops the new one and re-registers the prior.
    """

    action: str  # "reused" | "registered" | "service_swap"
    prior_collector: Optional["SimsysProcessCollector"] = None


class SimsysProcessCollector:
    """Emits simsys_process_cpu_seconds_total / memory_bytes / open_fds."""

    def __init__(self, service: str) -> None:
        self._service = service
        self._proc = psutil.Process()

    def collect(self):
        try:
            cpu = self._proc.cpu_times()
            cpu_total = float(cpu.user) + float(cpu.system)
        except (psutil.Error, OSError):
            cpu_total = 0.0

        cpu_family = CounterMetricFamily(
            "simsys_process_cpu_seconds_total",
            "Process CPU seconds (user + system) consumed since process start.",
            labels=["service"],
        )
        cpu_family.add_metric([self._service], cpu_total)
        yield cpu_family

        try:
            mem = self._proc.memory_info()
            rss = float(mem.rss)
            vms = float(mem.vms)
        except (psutil.Error, OSError):
            rss = 0.0
            vms = 0.0

        mem_family = GaugeMetricFamily(
            "simsys_process_memory_bytes",
            "Process memory in bytes. type=rss is resident set; type=vms is virtual.",
            labels=["service", "type"],
        )
        mem_family.add_metric([self._service, "rss"], rss)
        mem_family.add_metric([self._service, "vms"], vms)
        yield mem_family

        try:
            num_fds = float(self._proc.num_fds())
        except (psutil.Error, OSError, AttributeError):
            num_fds = 0.0

        fds_family = GaugeMetricFamily(
            "simsys_process_open_fds",
            "Open file descriptors for this process.",
            labels=["service"],
        )
        fds_family.add_metric([self._service], num_fds)
        yield fds_family


def register_process_collector(
    service: str,
) -> tuple["SimsysProcessCollector", ProcessCollectorRollbackState]:
    """Register (or re-use) the singleton process collector for ``service``.

    Returns ``(collector, rollback_state)`` where ``rollback_state``
    captures exactly what changed so adapter rollback can undo it
    precisely. See :class:`ProcessCollectorRollbackState` for the
    semantics of each action; install() rollback passes the state back
    into :func:`restore_process_collector` on failure.
    """
    global _collector
    with _lock:
        if _collector is not None:
            if _collector._service != service:
                # Service-swap: drop the prior collector but remember it
                # so rollback can re-register it if a later install step
                # fails. Without this capture, a swap-then-fail would
                # leave the PRIOR install's process metrics permanently
                # broken — its collector unregistered, no replacement.
                prior = _collector
                REGISTRY.unregister(prior)
                _collector = SimsysProcessCollector(service)
                REGISTRY.register(_collector)
                return _collector, ProcessCollectorRollbackState(
                    action="service_swap", prior_collector=prior
                )
            return _collector, ProcessCollectorRollbackState(action="reused")
        _collector = SimsysProcessCollector(service)
        REGISTRY.register(_collector)
        return _collector, ProcessCollectorRollbackState(action="registered")


def restore_process_collector(state: ProcessCollectorRollbackState) -> None:
    """Undo whatever register_process_collector() did, given its returned
    rollback state. Called from adapter install rollback when a later
    step fails.
    """
    global _collector
    if state.action == "reused":
        # Singleton was already correct — nothing to undo.
        return
    with _lock:
        if state.action == "registered":
            # We registered fresh. Drop everything.
            if _collector is not None:
                try:
                    REGISTRY.unregister(_collector)
                except KeyError:
                    pass
                _collector = None
            return
        if state.action == "service_swap":
            # Drop the new collector and re-register the prior one so
            # the previously-installed service's process metrics keep
            # flowing. If `register` is rejected for an already-
            # registered name (shouldn't happen — we just unregistered
            # the new one), swallow it; the prior collector still owns
            # its registration.
            if _collector is not None:
                try:
                    REGISTRY.unregister(_collector)
                except KeyError:
                    pass
            if state.prior_collector is not None:
                try:
                    REGISTRY.register(state.prior_collector)
                except ValueError:
                    pass
                _collector = state.prior_collector
            else:
                _collector = None
            return


def unregister_process_collector() -> None:
    """Test-only helper: drop the registered process collector."""
    global _collector
    with _lock:
        if _collector is not None:
            try:
                REGISTRY.unregister(_collector)
            except KeyError:
                pass
            _collector = None

File: node/src/process.ts
/**
 * Custom simsys_process_* collector.
 *
 * prom-client's default metrics cover most of this already (`process_*`,
 * `nodejs_*`) but they don't carry the `simsys_` prefix or the `service`
 * label. This module emits our own family alongside so cross-service PromQL
 * like `sum by (service) (simsys_process_memory_bytes{type="rss"})` works
 * unmodified across every app.
 *
 * Matches `simsys_metrics._process` (Python).
 *
 * State (registered flag, current service, the four metric refs, the
 * cumulative-CPU memo, and the per-collect mutex) lives on `globalThis`
 * so bundler chunk-splitting can't give each chunk its own copy. See
 * registry.ts header for the full rationale.
 */

import { readdirSync } from "node:fs";
import { Counter, Gauge } from "prom-client";
import { registry } from "./registry.js";

interface SimsysProcessState {
  registered: boolean;
  service: string | null;
  cpuTotal: Counter | null;
  memoryBytes: Gauge | null;
  openFds: Gauge | null;
  uptimeSeconds: Gauge | null;
  // Last observed cumulative CPU seconds — used to inc() the Counter by the
  // delta on each collect() cycle, preserving monotonic counter semantics so
  // rate() and reset-detection behave correctly.
  //
  // Updated atomically inside cpuCollectMutex below: prom-client's
  // Registry.metrics() walks collectors via Promise.all with no per-registry
  // mutex, so two concurrent scrapes (e.g. Prometheus + a sidecar push
  // monitor) can race the read-and-update sequence. Without serialization,
  // scrape A reads lastCpuSeconds=0, scrape B reads lastCpuSeconds=0, both
  // compute the full delta-from-zero and inc by ~total_cpu — counter
  // advances by 2× actual on every concurrent-scrape pair.
  lastCpuSeconds: number;
  // Globally-scoped serialization for the cpu collect() callback. Each new
  // invocation chains onto the previous one; the read-update of
  // lastCpuSeconds + the Counter.inc both run inside the chain.
  cpuCollectMutex: Promise<void>;
}

declare global {
  // eslint-disable-next-line no-var
  var __simsysMetricsProcessState: SimsysProcessState | undefined;
}

const _state: SimsysProcessState = (globalThis.__simsysMetricsProcessState ??= {
  registered: false,
  service: null,
  cpuTotal: null,
  memoryBytes: null,
  openFds: null,
  uptimeSeconds: null,
  lastCpuSeconds: 0,
  cpuCollectMutex: Promise.resolve(),
});

/**
 * Snapshot of process-collector state immediately before
 * registerProcessCollector() mutated it. Adapter rollback passes this
 * back to restoreProcessCollector() so a swap-then-fail sequence
 * doesn't leave a prior service's process metrics broken.
 *
 * `action` describes what registerProcessCollector did:
 *   - "reused":        collector already existed for this service; nothing changed.
 *   - "registered":    no collector existed; we registered fresh metrics. Rollback drops them.
 *   - "service-swap":  collector existed but for a different service; we relabelled it. Rollback restores priorService.
 */
export type ProcessCollectorRollbackState =
  | { action: "reused" }
  | { action: "registered" }
  | { action: "service-swap"; priorService: string };

function countOpenFds(): number {
  // Linux-only: /proc/self/fd lists one entry per open FD.
  if (process.platform !== "linux") return 0;
  try {
    return readdirSync("/proc/self/fd").length;
  } catch {
    return 0;
  }
}

export function registerProcessCollector(
  svc: string,
): ProcessCollectorRollbackState {
  if (_state.registered) {
    if (_state.service === svc) {
      return { action: "reused" };
    }
    // Re-install with a different service: refresh the static label so
    // future collect() callbacks tag samples with the new service.
    //
    // Drop ONLY the prior service's labelsets via .remove(...) — NOT
    // .reset() the whole metric. Resetting zeros the Counter for every
    // labelset (Prometheus reads it as a counter reset and gaps the
    // rate() series); .remove(prior) drops just the swapped-out
    // labelset so the live service's accumulator stays monotonic.
    //
    // Also: do NOT zero lastCpuSeconds. The cumulative process CPU is a
    // process-level value; preserving lastCpuSeconds means the next
    // collect() inc()s the new service's counter by a small delta from
    // the current cumulative, not by the full cumulative-from-zero
    // (which would create a one-scrape spike artifact).
    const priorService = _state.service ?? "";
    _state.service = svc;
    _state.cpuTotal?.remove({ service: priorService });
    _state.memoryBytes?.remove({ service: priorService, type: "rss" });
    _state.memoryBytes?.remove({ service: priorService, type: "heapUsed" });
    _state.memoryBytes?.remove({ service: priorService, type: "heapTotal" });
    _state.memoryBytes?.remove({ service: priorService, type: "external" });
    _state.openFds?.remove({ service: priorService });
    _state.uptimeSeconds?.remove({ service: priorService });
    return { action: "service-swap", priorService };
  }
  _state.service = svc;
  _state.registered = true;

  _state.cpuTotal = new Counter({
    name: "simsys_process_cpu_seconds_total",
    help: "Process CPU seconds (user + system) consumed since process start.",
    labelNames: ["service"],
    registers: [registry],
    async collect() {
      // process.cpuUsage() returns microseconds since process start, which
      // is monotonic. Inc by the delta since the last collect so the
      // prom-client Counter's internal value stays monotonic too — using
      // reset()+inc() would defeat prom-client's reset-detection (rate()
      // over a scrape boundary would behave undefined).
      //
      // Serialize via cpuCollectMutex so concurrent /metrics scrapes
      // can't both read the same lastCpuSeconds and double-count.
      const counter = this;
      const next = _state.cpuCollectMutex.then(() => {
        const { user, system } = process.cpuUsage();
        const total = (user + system) / 1_000_000;
        const delta = Math.max(0, total - _state.lastCpuSeconds);
        _state.lastCpuSeconds = total;
        if (delta > 0) {
          counter.inc({ service: _state.service! }, delta);
        }
      });
      _state.cpuCollectMutex = next.catch(() => undefined); // swallow for the chain
      await next;
    },
  });

  _state.memoryBytes = new Gauge({
    name: "simsys_process_memory_bytes",
    help: "Process memory in bytes. type=rss is resident set; type=heapUsed is V8 heap used; type=heapTotal is V8 heap total.",
    labelNames: ["service", "type"],
    registers: [registry],
    collect() {
      const mem = process.memoryUsage();
      this.set({ service: _state.service!, type: "rss" }, mem.rss);
      this.set({ service: _state.service!, type: "heapUsed" }, mem.heapUsed);
      this.set({ service: _state.service!, type: "heapTotal" }, mem.heapTotal);
      this.set({ service: _state.service!, type: "external" }, mem.external);
    },
  });

  _state.openFds = new Gauge({
    name: "simsys_process_open_fds",
    help: "Open file descriptors for this process (Linux only; 0 elsewhere).",
    labelNames: ["service"],
    registers: [registry],
    collect() {
      this.set({ service: _state.service! }, countOpenFds());
    },
  });

  _state.uptimeSeconds = new Gauge({
    name: "simsys_process_uptime_seconds",
    help: "Process uptime in seconds.",
    labelNames: ["service"],
    registers: [registry],
    collect() {
      this.set({ service: _state.service! }, process.uptime());
    },
  });

  return { action: "registered" };
}

/**
 * Undo whatever registerProcessCollector() did, given its returned state.
 * Called from adapter install rollback when a later step fails, so a
 * service-swap doesn't leave the PRIOR install's process metrics broken
 * (label rewritten to the failed install's service) and a fresh
 * registration doesn't leave dangling collectors in the registry.
 */
export function restoreProcessCollector(
  state: ProcessCollectorRollbackState,
): void {
  switch (state.action) {
    case "reused":
      // Nothing changed; nothing to undo.
      return;
    case "service-swap":
      // We mutated the static `service` global in place — restore it.
      // Drop the failed install's service labelset (the symmetric
      // counterpart to the swap-time .remove of the prior service)
      // via .remove(...), NOT .reset(). Reset would zero every
      // labelset including the prior one Prometheus has been scraping,
      // creating a counter-reset artifact in rate() — .remove drops
      // just the failed install's labelset and leaves the prior
      // service's accumulator intact and monotonic.
      //
      // Do NOT zero lastCpuSeconds: the prior service's collect()
      // resumes from the current cumulative, attributing only the
      // delta-since-prior-collect to it. Resetting would make the
      // very first post-rollback inc() jump the counter by the full
      // process-cumulative CPU.
      {
        const failedService = _state.service ?? "";
        _state.service = state.priorService;
        if (failedService && failedService !== state.priorService) {
          _state.cpuTotal?.remove({ service: failedService });
          _state.memoryBytes?.remove({ service: failedService, type: "rss" });
          _state.memoryBytes?.remove({ service: failedService, type: "heapUsed" });
          _state.memoryBytes?.remove({ service: failedService, type: "heapTotal" });
          _state.memoryBytes?.remove({ service: failedService, type: "external" });
          _state.openFds?.remove({ service: failedService });
          _state.uptimeSeconds?.remove({ service: failedService });
        }
      }
      return;
    case "registered":
      // We registered fresh — drop everything.
      _unregisterAll();
      return;
  }
}

function _unregisterAll(): void {
  _state.registered = false;
  _state.service = null;
  _state.lastCpuSeconds = 0;
  _state.cpuCollectMutex = Promise.resolve();
  registry.removeSingleMetric("simsys_process_cpu_seconds_total");
  registry.removeSingleMetric("simsys_process_memory_bytes");
  registry.removeSingleMetric("simsys_process_open_fds");
  registry.removeSingleMetric("simsys_process_uptime_seconds");
  _state.cpuTotal = _state.memoryBytes = _state.openFds = _state.uptimeSeconds = null;
}

export function _resetForTests(): void {
  // Best-effort test helper.
  _unregisterAll();
}
